### PR TITLE
Bio.SearchIO parser for InterProScan XML

### PR DESCRIPTION
**Pull request by azneto**
This pull request addresses issue #1536 

I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

I am happy be thanked by name in the ``NEWS.rst`` and ``CONTRIB.rst`` files,
and have added myself to those files as part of this pull request. (*This
acknowledgement is optional. Note we list the names sorted alphabetically.*)

### CODE REVIEW

File: Tests/test_SearchIO_interpro_xml.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+TEST_DIR = 'InterProScan'
+FMT = 'interpro-xml'
+
+
+def get_file(filename):
+    """Returns the path of a test file."""
+    return os.path.join(TEST_DIR, filename)
+
+
+class BlastnXmlCases(unittest.TestCase):
```
Comment (bow) [inline Tests/test_SearchIO_interpro_xml.py]: This should be `InteproscanXmlCases`, I believe :).

File: Bio/SearchIO/InterproIO/__init__.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+"""InterProScan."""
+from .interpro_xml import InterproXmlParser
+
+
```
Comment (bow) [inline Bio/SearchIO/InterproIO/__init__.py]: Would it also be useful to add more docstrings here about the format in general? For example, you can say which InterProScan output type is supported or how to access certain values. The docstrings in `BlastIO/__init__.py` and/or `HmmerIO/__init__.py` may be useful as a template.

File: Bio/SearchIO/__init__.py [hunk 3]
```diff
@@ -232,6 +232,7 @@
         'hmmscan3-domtab': ('HmmerIO', 'Hmmer3DomtabHmmhitParser'),
         'hmmsearch3-domtab': ('HmmerIO', 'Hmmer3DomtabHmmqueryParser'),
         'phmmer3-domtab': ('HmmerIO', 'Hmmer3DomtabHmmqueryParser'),
+        'interproscan-xml': ('InterproscanIO', 'InterproscanXmlParser'),
```
Comment (peterjc) [inline Bio/SearchIO/__init__.py]: The dictionary keys seem to be alphabetical here.

**Comment by peterjc** [comment]
My query to @bow is about the new attributes being added, and if this the best solution long term? Should we favour a dictionary approach over adding new format specific attributes?:

```
 +        self.type = None
 +        self.target = None
 +        self.target_version = None
 +        self.dbxrefs = []
```

We have the precedent of ``.dbxrefs`` in the ``Seq*`` objects so the naming of that is fine.

File: Tests/test_SearchIO_interpro_xml.py (hunk 1 above)
Comment (azneto) [reply Tests/test_SearchIO_interpro_xml.py]: done

File: Bio/SearchIO/InterproIO/__init__.py (hunk 2 above)
Comment (azneto) [reply Bio/SearchIO/InterproIO/__init__.py]: done

File: Bio/SearchIO/__init__.py (hunk 3 above)
Comment (azneto) [reply Bio/SearchIO/__init__.py]: done

**Comment by azneto** [comment]
@bow, it's good to know it's in the right track.
I'll address each topic:

1. matchTypes and locationTypes have always the same prefix. So I've decided to store the string excluding the '-match' or '-location'. The variable is hit.type, but I'd like to hear from you about Peter's questioning. I'm not quite sure about how to implement it in another way (dict).
2. InterProScan does not store 'no hits' and I didn't find flags to modify the xml output. 
3. Nice observation. I've just changed every mention to the software to 'interproscan'.

**Comment by bow** [comment]
@peterjc Hmm, a dictionary sound more preferable. This way the `__str__` display can be made generic for other format attributes as well.

@azneto For the dict, we can have one new attribute (e.g. `format_attributes`, which is of type `dict`). Format-specific parsers could then add their own attributes there.

For your responses:

1. In that case, related to the current naming of the types also: `type` is perhaps better replaced with a synonym (`kind`? `database_type`? `search_type`? `hit_type`?). This is because `type` is a Python keyword and it's generally a good idea to avoid shadowing the builtin keywords.

2 & 3. Ok.

**Comment by azneto** [comment]
I think I've got it.

**Comment by azneto** [comment]
Hi @bow, did you have the chance to review it?

**Comment by bow** [comment]
@azneto ~ Ah, I missed the notification for the last changeset, sorry. I seems ready for merging now. @peterjc, anything else that should be added?

**Comment by peterjc** [comment]
From a usability perspective, is ``.format_attributes`` as the dict name the best choice? Is ``.attributes`` enough? Or maybe ``.qualifiers`` or ``.annotations`` better to match other usage in Biopython?

**Comment by azneto** [comment]
Such name refers to format specific attributes. But I agree with you, .attributes will definitely improve usability.

**Comment by peterjc** [comment]
I think this would deserve to be mentioned in the ``NEWS.rst`` file, along with adding your name for the release and to the ``CONTRIB.rst`` file.

Also once this is merged, we'd want to add an entry to the website listing the new format in Biopython 1.71:
http://biopython.org/wiki/SearchIO
https://github.com/biopython/biopython.github.io/edit/master/wiki/SearchIO.md

**Comment by azneto** [comment]
Sounds good. My name is already in the CONTRIB.rst and NEWS.rst files. We just need to add the information about the tool.

**Comment by bow** [comment]
Great! If there are no further objections, I will merge this today, no sooner than 12 noon CET. Thank you @azneto and for @peterjc for reviewing.

**Comment by peterjc** [comment]
Is it easy to add indexing as well as a future addition? Or is there little need for it?

**Comment by bow** [comment]
There is some resemblance to the `blast-xml` structure here, so I'm guessing the approaches we use in `blast-xml` indexing can be used if we do need indexing.

But for now, I guess this can be merged as-is (unless @azneto wants to work on indexing before this is merged to `master`?). This already covers the majority of the use cases it seems :).

**Comment by peterjc** [comment]
Sure, merge as is - indexing can be added later.

**Comment by azneto** [comment]
I think it might be useful add indexing, but I never worked with this and I'll need time to study it. We can add that later on.

**Comment by bow** [comment]
Great, ok, I'm merging this now then :+1:. Thanks!

**Pull request merged**